File: ApiSuvesaPos/Negocio/Logica/Ubicaciones.vb
Imports Negocio.Utilidades
Namespace Logica
    Public Class Ubicaciones

        Private db As Datos.Class.Ubicaciones

        Sub New()
            Me.db = New Datos.Class.Ubicaciones
        End Sub

        Public Function Buscar(pornombre As Boolean, filtro As String) As List(Of Datos.Models.Ubicacione)
            Dim datos As New List(Of Datos.Models.Ubicacione)
            datos = Me.db.Buscar(pornombre, filtro)

            Dim Resultado As New List(Of Datos.Models.Ubicacione)
            For Each ubi As Datos.Models.Ubicacione In datos
                Dim Ubicacion As New Datos.Models.Ubicacione
                Ubicacion.Codigo = ubi.Codigo
                Ubicacion.Descripcion = ubi.Descripcion
                Ubicacion.Observaciones = ubi.Observaciones
                Ubicacion.Activa = ubi.Activa

                For Each subUbi As Datos.Models.SubUbicacion In Me.db.BuscarDetalle(ubi.Codigo)
                    Dim SubUbicacion As New Datos.Models.SubUbicacion
                    SubUbicacion.CodUbicacion = subUbi.CodUbicacion
                    SubUbicacion.CodSubUbicacion = subUbi.CodSubUbicacion
                    SubUbicacion.Codigo = subUbi.Codigo
                    SubUbicacion.DescripcionD = subUbi.DescripcionD
                    SubUbicacion.Observaciones = subUbi.Observaciones
                    Ubicacion.SubUbicacions.Add(SubUbicacion)
                Next

                Resultado.Add(Ubicacion)
            Next

            Return Resultado
        End Function

        Public Function Crear(ubicacion As Datos.Models.Ubicacione) As String
            Return Me.db.Crear(ubicacion)
        End Function

        Public Function Editar(id As Decimal, ubicacion As Datos.Models.Ubicacione) As String
            Dim res As String = Me.db.Editar(id, ubicacion)
            If res = "0" Then
                Return "No existe el valor"
            Else
                Return res
            End If
        End Function

        Public Function Eliminar(id As Decimal) As String
            Dim res As String = Me.db.Borrar(id)
            If res = "0" Then
                Return "No existe el valor"
            Else
                Return res
            End If
        End Function

    End Class

End Namespace

File: ApiSuvesaPos/Datos/Class/Ubicaciones.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Datos.Models;
using Microsoft.EntityFrameworkCore;

namespace Datos.Class
{
	public class Ubicaciones
	{
		private SeePOSContext entities;

		public Ubicaciones()
		{
			entities = new SeePOSContext();
		}

		public int Crear(Models.Ubicacione ubicaciones) // registro de Ubicaciones
		{
			try
			{
				entities.Ubicaciones.Add(ubicaciones);
				return entities.SaveChanges();
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

		public List<Models.SubUbicacion> BuscarDetalle(long Id)  //consultar Ubicaciones
		{
			try
			{
				List<Models.SubUbicacion> result;
				
					var temp = from c in entities.SubUbicacions
							   where c.CodUbicacion == Id
							   select c;
					result = temp.ToList<Models.SubUbicacion>();
				
				if (result.Count > 0)
				{
					return result;
				}
				else
				{
					return result = null;
				}
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

		public List<Models.Ubicacione> Buscar(bool porNombre, string filtro)  //consultar Ubicaciones
		{
			try
			{
				List<Models.Ubicacione> result;
				if (porNombre == true)
				{
					var temp = from c in entities.Ubicaciones
							   where c.Descripcion.Contains(filtro) 
							   select c;
					result = temp.ToList<Models.Ubicacione>();
				}
				else
				{
					var temp = from c in entities.Ubicaciones
							   where c.Codigo == Convert.ToInt32(filtro)
							   select c;
					result = temp.ToList<Models.Ubicacione>();
				}
				if (result.Count > 0)
				{
					return result;
				}
				else
				{
					return result = null;
				}
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

		public int Editar(long id, Models.Ubicacione Ubicacion) //editar Ubicaciones
		{
			try
			{
				var p = entities.Ubicaciones.Find(id);
				Models.Ubicacione Ubicaciones = p;
				if (Ubicaciones != null)
				{
					Ubicaciones.Codigo = Ubicacion.Codigo;
					Ubicaciones.Descripcion = Ubicacion.Descripcion;
					Ubicaciones.Observaciones = Ubicacion.Observaciones;
					Ubicaciones.Activa = Ubicacion.Activa;

					Models.SubUbicacion nuevaLinea;
					foreach (Models.SubUbicacion Detalle in Ubicacion.SubUbicacions)
					{
						//Agrega nuevos registros
						if (Detalle.Codigo == "0")
						{
							nuevaLinea = new Models.SubUbicacion();
							nuevaLinea.CodUbicacion = Detalle.CodUbicacion;
							nuevaLinea.CodSubUbicacion = Detalle.CodSubUbicacion;
							nuevaLinea.DescripcionD = Detalle.DescripcionD;
							nuevaLinea.Observaciones = Detalle.Observaciones;
							Ubicaciones.SubUbicacions.Add(nuevaLinea);
						}
						else
						{
							//Actualiza los detalles
							var a = entities.SubUbicacions.Find(Detalle.Codigo);
							Models.SubUbicacion lineaModificada = a;
							if (lineaModificada != null)
							{
								lineaModificada.CodUbicacion = Detalle.CodUbicacion;
								lineaModificada.CodSubUbicacion = Detalle.CodSubUbicacion;
								lineaModificada.Codigo = Detalle.Codigo;
								lineaModificada.DescripcionD = Detalle.DescripcionD;
								lineaModificada.Observaciones = Detalle.Observaciones;

								entities.Entry(lineaModificada).State = EntityState.Modified;
								entities.SaveChanges();
							}
						}

					}

					entities.Entry(Ubicaciones).State = EntityState.Modified;
					return entities.SaveChanges();
				}
				else
				{
					return 0;// no se encotro el registro solicitado.
				}
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

		public int Borrar(long id) //borrar Ubicaciones
		{
			try
			{
				var p = entities.Ubicaciones.Find(id);
				Models.Ubicacione Nuevo = p;
				if (Nuevo != null)
				{
					entities.Remove(p);
					return entities.SaveChanges();
				}
				else
				{
					return 0;// no se encotro el registro solicitado.
				}
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}
	}
}


